packages/lxd/src/query.coffee.md

```diff
@@ -36,16 +36,49 @@ Flags:
             The API path in the form of `[<remote>:]<API path>`, for example
             `/1.0/instances/c1`
             '''
+          'wait':
+            type: 'boolean'
+            default: false
+            description: '''
+            If true, activates the wait flag that waits for the operation to complete.
+            '''
+          'request':
+            enum: ['GET', 'PUT', 'DELETE', 'POST', 'PATCH']
+            default: 'GET'
+            description: '''
+            Action to use for the API call.
+            '''
+          'data':
+            type: 'string'
+            description: '''
+            Data to send to the action in the form of application/json stringified object.
+            '''
+          'parse':
+            type: 'boolean'
+            default: true
+            description: '''
+            Parse the response of the request as application/json.
+            ''' 
 
 ## Handler
 
     handler = ({config}) ->
       {stdout} = await @execute
-        command: [
-          'lxc', 'query', config.path
-        ].join ' '
+        command: """
+        #{[
+          'lxc', 'query', 
+          "--wait" if config.wait, 
+          "--request", config.request, 
+          "--data '#{config.data}'" if config.data?, 
+          config.path
+        ].join ' '}
+        """
       $status: true
-      data: JSON.parse stdout
+      # note, the parse flag is used because some queries return a string and it would mean refactoring every query if the default wasn't set to true.
+      if config.parse
+        data: JSON.parse stdout
+      else
+        data: stdout
```

Comment:
Shouldn't just using the flag `--raw` force the response to always be JSON ?

Reply:
The `--raw` flag allows for the raw query to appear, that is to say for the status code. 

Example with `--raw` flag:

```bash
[kellian@kellian-dell:~]$ lxc query --raw /1.0/instances
{
	"type": "sync",
	"status": "Success",
	"status_code": 200,
	"operation": "",
	"error_code": 0,
	"error": "",
	"metadata": [
		"/1.0/instances/master-node",
		"/1.0/instances/worker-node-1",
		"/1.0/instances/worker-node-2",
		"/1.0/instances/worker-node-3",
		"/1.0/instances/worker-node-4"
	]
}
```

Example without flag:

```bash
[kellian@kellian-dell:~]$ lxc query /1.0/instances
[
	"/1.0/instances/master-node",
	"/1.0/instances/worker-node-1",
	"/1.0/instances/worker-node-2",
	"/1.0/instances/worker-node-3",
	"/1.0/instances/worker-node-4"
]
```

What the `parse` option does, is parse the response of a query, which can be either a JSON output or a string. For instance, if you try getting a file from an instance, you would do this:

```bash
[kellian@kellian-dell:~]$ lxc query /1.0/instances/master-node/files?path=/root/kubeadm/kubeadm-config.yaml
# kubeadm-config.yaml

kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta3
kubernetesVersion: "v1.23.6"
clusterName: "k8s-cluster"
networking:
  podSubnet: 192.168.0.0/24
```

The answer from this command is not a JSON file, but a text file. We can't parse it, else we would run into an error.

I hope this was clearer.

Reply:
Yes, thank you

Reply:
> Yes, thank you

I think it would be better to simply turn the flag to always false, and force the parsing on the features that need it. At least, it would be clearer than having to deselect parsing.

What do you think about this ? It would mean refactoring a lot, but it would be more appropriate.

Reply:
Agree, I dislike boolean default to `false`. Make it consistent with [`nikita.execute` format option](https://github.com/adaltas/node-nikita/blob/master/packages/core/src/actions/execute/index.coffee.md):

```
      'format':
        type: 'string'
        enum: ['json', 'yaml']
        description: '''
        Convert the stdout to a Javascript value or object.
        '''
```

Reply:
The thing is, he is not parsing the data to YAML, he's getting the content of a YAML file. But it could have been an XML file or something else. Therefore, I don't think an enum will help much in this case. Except if you want an enum with values: `json, yaml, raw` ?

Reply:
@kellianoy what happens when you get a binary file ? such as a png

Reply:
@gboutry Getting a png won't get you anywhere. I don't think there is a proper option to send images in [here](https://linuxcontainers.org/lxd/api/master/#/). 

Sample result of getting a png: 

```bash
�6��e��$�%�I.!p�杂�_�*w�j��������*�3�揭]���ݻT������������U}.E'���Ǯ;J��7��f� ^��/P��g��K?�w��[9��ŷe���U�˷ݿt˽�����uZ��+��7�Wup����w�z����]���!�_�_W2�'�.�;�,)u�f܎�9��������eNW'A�v�)f������B^,���(����2������9��S�
```